### PR TITLE
fix: Reenable concurrent hashing

### LINKED ISSUE #10872: Fix how concurrent hashing handles downstream backpressure
### Problem

- Under the current framework, concurrent task schedulers cannot properly handle backpressure from the next component in the pipeline
- If the component immediately following a concurrent scheduler applies backpressure, the concurrent scheduler will continue doing work and parking the result, unable to pass the unit of work along
- This results in a proliferation of paused threads

### Short Term Solution

- Create a supporting sequential scheduler immediately after the concurrent scheduler in the pipeline
- Create a shared `ObjectCounter` that will on-ramp when adding a task to the concurrent scheduler, and off-ramp when emitting the task from the supporting scheduler
- When backpressure is applied to the pair of schedulers from further along the pipeline, tasks that have been completed by the concurrent scheduler will accumulate in the supporting scheduler, and not result in a proliferation of paused threads
image

### Long Term Solution

- Bundle the above logic into the concurrent scheduler, so that it can be achieved via the normal task scheduler interface (https://github.com/hashgraph/hedera-services/issues/11063)

### PR DESCRIPTION
**Pull request by alittley**
- closes #10872
    - the longterm solution originally described in that issue has been split into a separate ticket https://github.com/hashgraph/hedera-services/issues/11063

- [normal diagram](https://mermaid.ink/svg/Zmxvd2NoYXJ0IExSCnYwKCgiQCIpKQpzdWJncmFwaCB2MVsiQ29uc2Vuc3VzIFBpcGVsaW5lIl0Kc3ViZ3JhcGggdjJbIkxpbmtlZCBFdmVudCBJbnRha2UiXQp2M3t7ImdldE5vbkFuY2llbnRFdmVudFdpbmRvdyJ9fQp2NFsibGlua2VkRXZlbnRJbnRha2U8YnIgLz7imIUiXQp2NXt7ImxpbmtlZEV2ZW50SW50YWtlU3BsaXR0ZXIifX0KZW5kCnY2WyJpbk9yZGVyTGlua2VyPGJyIC8+ypgiXQplbmQKc3ViZ3JhcGggdjdbIkludGFrZSBQaXBlbGluZSJdCnN1YmdyYXBoIHY4WyJFdmVudCBIYXNoaW5nIl0KdjlbWyJldmVudEhhc2hlciJdXQp2MTBbInBvc3RIYXNoQ29sbGVjdG9yIl0KZW5kCnN1YmdyYXBoIHYxMVsiRXZlbnQgVmFsaWRhdGlvbiJdCnYxMlsiZXZlbnREZWR1cGxpY2F0b3I8YnIgLz7KmCJdCnYxM1siZXZlbnRTaWduYXR1cmVWYWxpZGF0b3I8YnIgLz7KmCJdCnYxNFsiaW50ZXJuYWxFdmVudFZhbGlkYXRvcjxiciAvPuKAoCJdCmVuZAp2MTVbIk9ycGhhbiBCdWZmZXI8YnIgLz7KmCJdCmVuZApzdWJncmFwaCB2MTZbIlBDRVMiXQp2MTdbLyJldmVudER1cmFiaWxpdHlOZXh1cyIvXQp2MThbLyJwY2VzU2VxdWVuY2VyIi9dCnYxOVsicGNlc1dyaXRlcjxiciAvPkDKmCJdCmVuZApzdWJncmFwaCB2MjBbIlN0YXRlIEZpbGUgTWFuYWdlbWVudCJdCnYyMXt7ImV4dHJhY3RPbGRlc3RNaW5pbXVtR2VuZXJhdGlvbk9uRGlzayJ9fQp2MjJ7eyJzYXZlVG9EaXNrRmlsdGVyIn19CnYyM1sic2lnbmVkU3RhdGVGaWxlTWFuYWdlciJdCnYyNHt7InRvU3RhdGVXcml0dGVuVG9EaXNrQWN0aW9uIn19CmVuZApzdWJncmFwaCB2MjVbIlN0YXRlIFNpZ25hdHVyZSBDb2xsZWN0aW9uIl0KdjI2e3siYWxsU3RhdGVzUmVzZXJ2ZXIifX0KdjI3e3siY29tcGxldGVTdGF0ZUZpbHRlciJ9fQp2Mjh7eyJjb21wbGV0ZVN0YXRlc1Jlc2VydmVyIn19CnYyOXt7ImV4dHJhY3RDb25zZW5zdXNTaWduYXR1cmVUcmFuc2FjdGlvbnMifX0KdjMwe3siZXh0cmFjdFByZWNvbnNlbnN1c1NpZ25hdHVyZVRyYW5zYWN0aW9ucyJ9fQp2MzF7eyJyZXNlcnZlZFN0YXRlU3BsaXR0ZXIifX0KdjMyWyJzdGF0ZVNpZ25hdHVyZUNvbGxlY3RvciJdCmVuZAp2MzNbWyJhcHBsaWNhdGlvblRyYW5zYWN0aW9uUHJlaGFuZGxlciJdXQp2MzRbImV2ZW50Q3JlYXRpb25NYW5hZ2VyPGJyIC8+ypjimaEiXQp2MzVbImhlYXJ0YmVhdCJdCnYzNlsvInBjZXNSZXBsYXllciIvXQp2MzdbInN0YXRlU2lnbmVyIl0KdjM4KCgiypgiKSkKdjM5KCgi4oCgIikpCnY0MCgoIuKYhSIpKQp2NDEoKCLimaEiKSkKdjE1IC0tICJwcmVjb25zZW5zdXMgZXZlbnRzIiAtLT4gdjMzCnYxNSAtLSAicG9zc2libGUgcGFyZW50cyIgLS0+IHYzNAp2MTUgLS0gInByZWNvbnNlbnN1cyBldmVudHMiIC0tPiB2MzAKdjE1IC0tICJldmVudHMgdG8gc2VxdWVuY2UiIC0tPiB2MTgKdjI2IC0tICJzdGF0ZXMiIC0tPiB2MjcKdjI2IC0tICJzdGF0ZXMiIC0tPiB2MjIKdjMzIC0uICJmdXR1cmVzIiAuLW8gdjQwCnYyNyAtLT4gdjI4CnYzNCAtLiAibm9uLXZhbGlkYXRlZCBldmVudHMiIC4tPiB2MzkKdjEyIC0tICJldmVudHMgd2l0aCB1bnZhbGlkYXRlZCBzaWduYXR1cmVzIiAtLT4gdjEzCnY5IC0tICJoYXNoZWQgZXZlbnRzIiAtLT4gdjEwCnYxMyAtLSAidW5vcmRlcmVkIGV2ZW50cyIgLS0+IHYxNQp2MjkgLS0gImNvbnNlbnN1cyBzaWduYXR1cmUgdHJhbnNhY3Rpb25zIiAtLT4gdjMyCnYyMSAtLSAibWluaW11bSBpZGVudGlmaWVyIHRvIHN0b3JlIiAtLT4gdjE5CnYzMCAtLSAicHJlY29uc2Vuc3VzIHNpZ25hdHVyZSB0cmFuc2FjdGlvbnMiIC0tPiB2MzIKdjMgLS4gIm5vbi1hbmNpZW50IGV2ZW50IHdpbmRvdyIgLi0+IHYzOAp2MzUgLS0gImhlYXJ0YmVhdCIgLS0+IHY0MQp2NiAtLSAibGlua2VkIGV2ZW50cyIgLS0+IHY0CnYxNCAtLSAibm9uLWRlZHVwbGljYXRlZCBldmVudHMiIC0tPiB2MTIKdjQgLS0gInJvdW5kIGxpc3RzIiAtLT4gdjUKdjQgLS0gImZsdXNoIHJlcXVlc3QiIC0tPiB2MTkKdjUgLS0gInJvdW5kcyIgLS0+IHYzCnYzNiAtLSAiZG9uZSBzdHJlYW1pbmcgcGNlcyIgLS0+IHYwCnYzNiAtLSAiZXZlbnRzIHRvIGhhc2giIC0tPiB2OQp2MTggLS0gInVubGlua2VkIGV2ZW50cyIgLS0+IHY2CnYxOCAtLSAiZXZlbnRzIHRvIHdyaXRlIiAtLT4gdjE5CnYxOSAtLSAibGF0ZXN0IGR1cmFibGUgc2VxdWVuY2UgbnVtYmVyIiAtLT4gdjE3CnYxMCAtLSAibm9uLXZhbGlkYXRlZCBldmVudHMiIC0tPiB2MTQKdjMxIC0tPiB2MjYKdjIyIC0tICJzdGF0ZXMgdG8gc2F2ZSIgLS0+IHYyMwp2MjMgLS0gInN0YXRlIHNhdmluZyByZXN1bHQiIC0tPiB2MjEKdjIzIC0tICJzdGF0ZSBzYXZpbmcgcmVzdWx0IiAtLT4gdjI0CnYzMiAtLSAicmVzZXJ2ZWQgc3RhdGUgbGlzdHMiIC0tPiB2MzEKY2xhc3NEZWYgczAgZmlsbDojZjg4LHN0cm9rZTojMDAwLHN0cm9rZS13aWR0aDoycHgKY2xhc3MgdjAsdjM4LHYzOSx2NDAsdjQxIHMwCmNsYXNzRGVmIHMxIGZpbGw6IzlDRixzdHJva2U6IzAwMCxzdHJva2Utd2lkdGg6MnB4CmNsYXNzIHYxLHYxNix2MjAsdjI1LHY3IHMxCmNsYXNzRGVmIHMyIGZpbGw6I0JFRixzdHJva2U6IzAwMCxzdHJva2Utd2lkdGg6MnB4CmNsYXNzIHYxMSx2Mix2OCBzMgpjbGFzc0RlZiBzMyBmaWxsOiNjY2Msc3Ryb2tlOiMwMDAsc3Ryb2tlLXdpZHRoOjJweApjbGFzcyB2MTcsdjE4LHYyMSx2MjIsdjI0LHYyNix2MjcsdjI4LHYyOSx2Myx2MzAsdjMxLHYzNix2NSBzMwpjbGFzc0RlZiBzNCBmaWxsOiNmZjksc3Ryb2tlOiMwMDAsc3Ryb2tlLXdpZHRoOjJweApjbGFzcyB2MTAsdjEyLHYxMyx2MTQsdjE1LHYxOSx2MjMsdjMyLHYzMyx2MzQsdjM1LHYzNyx2NCx2Nix2OSBzNAo=?bgColor=e8e8e8)
- [concise diagram](https://mermaid.ink/svg/Zmxvd2NoYXJ0IExSCnYwKCgiQCIpKQpzdWJncmFwaCB2MVsiQ29uc2Vuc3VzIFBpcGVsaW5lIl0KdjJbIkxpbmtlZCBFdmVudCBJbnRha2U8YnIgLz7imIUiXQp2M1siaW5PcmRlckxpbmtlcjxiciAvPsqYIl0KZW5kCnN1YmdyYXBoIHY0WyJJbnRha2UgUGlwZWxpbmUiXQp2NVtbIkV2ZW50IEhhc2hpbmciXV0Kc3ViZ3JhcGggdjZbIkV2ZW50IFZhbGlkYXRpb24iXQp2N1siZXZlbnREZWR1cGxpY2F0b3I8YnIgLz7KmCJdCnY4WyJldmVudFNpZ25hdHVyZVZhbGlkYXRvcjxiciAvPsqYIl0KdjlbImludGVybmFsRXZlbnRWYWxpZGF0b3I8YnIgLz7igKAiXQplbmQKdjEwWyJPcnBoYW4gQnVmZmVyPGJyIC8+ypgiXQplbmQKdjExWyJQQ0VTIFdyaXRlcjxiciAvPkBzypgiXQp2MTJbIlN0YXRlIEZpbGUgTWFuYWdlbWVudCJdCnYxM1siU3RhdGUgU2lnbmF0dXJlIENvbGxlY3Rpb24iXQp2MTRbWyJhcHBsaWNhdGlvblRyYW5zYWN0aW9uUHJlaGFuZGxlciJdXQp2MTVbImV2ZW50Q3JlYXRpb25NYW5hZ2VyPGJyIC8+ypjimaEiXQp2MTZbImhlYXJ0YmVhdCJdCnYxN1svInBjZXNSZXBsYXllciIvXQp2MTgoKCJzIikpCnYxOVsic3RhdGVTaWduZXIiXQp2MjAoKCLKmCIpKQp2MjEoKCLigKAiKSkKdjIyKCgi4piFIikpCnYyMygoIuKZoSIpKQp2NSAtLSAibm9uLXZhbGlkYXRlZCBldmVudHMiIC0tPiB2OQp2MiAtLSAiZmx1c2ggcmVxdWVzdCIgLS0+IHYxMQp2MiAtLiAibm9uLWFuY2llbnQgZXZlbnQgd2luZG93IiAuLT4gdjIwCnYxMCAtLSAiZXZlbnRzIHRvIHNlcXVlbmNlIiAtLT4gdjExCnYxMCAtLSAicHJlY29uc2Vuc3VzIGV2ZW50cyIgLS0+IHYxMwp2MTAgLS0gInByZWNvbnNlbnN1cyBldmVudHMiIC0tPiB2MTQKdjEwIC0tICJwb3NzaWJsZSBwYXJlbnRzIiAtLT4gdjE1CnYxMSAtLSAidW5saW5rZWQgZXZlbnRzIiAtLT4gdjMKdjEyIC0tICJtaW5pbXVtIGlkZW50aWZpZXIgdG8gc3RvcmUiIC0tPiB2MTgKdjEzIC0tICJzdGF0ZXMiIC0tPiB2MTIKdjE0IC0uICJmdXR1cmVzIiAuLW8gdjIyCnYxNSAtLiAibm9uLXZhbGlkYXRlZCBldmVudHMiIC4tPiB2MjEKdjcgLS0gImV2ZW50cyB3aXRoIHVudmFsaWRhdGVkIHNpZ25hdHVyZXMiIC0tPiB2OAp2OCAtLSAidW5vcmRlcmVkIGV2ZW50cyIgLS0+IHYxMAp2MTYgLS0gImhlYXJ0YmVhdCIgLS0+IHYyMwp2MyAtLSAibGlua2VkIGV2ZW50cyIgLS0+IHYyCnY5IC0tICJub24tZGVkdXBsaWNhdGVkIGV2ZW50cyIgLS0+IHY3CnYxNyAtLSAiZG9uZSBzdHJlYW1pbmcgcGNlcyIgLS0+IHYwCnYxNyAtLSAiZXZlbnRzIHRvIGhhc2giIC0tPiB2NQpjbGFzc0RlZiBzMCBmaWxsOiNmODgsc3Ryb2tlOiMwMDAsc3Ryb2tlLXdpZHRoOjJweApjbGFzcyB2MCx2MTgsdjIwLHYyMSx2MjIsdjIzIHMwCmNsYXNzRGVmIHMxIGZpbGw6IzlDRixzdHJva2U6IzAwMCxzdHJva2Utd2lkdGg6MnB4CmNsYXNzIHYxLHY0IHMxCmNsYXNzRGVmIHMyIGZpbGw6I2ZmOSxzdHJva2U6IzAwMCxzdHJva2Utd2lkdGg6MnB4CmNsYXNzIHYxMCx2MTEsdjEyLHYxMyx2MTQsdjE1LHYxNix2MTksdjIsdjMsdjUsdjcsdjgsdjkgczIKY2xhc3NEZWYgczMgZmlsbDojQkVGLHN0cm9rZTojMDAwLHN0cm9rZS13aWR0aDoycHgKY2xhc3MgdjYgczMKY2xhc3NEZWYgczQgZmlsbDojY2NjLHN0cm9rZTojMDAwLHN0cm9rZS13aWR0aDoycHgKY2xhc3MgdjE3IHM0Cg==?bgColor=e8e8e8)

### CODE REVIEW

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/wiring/PlatformSchedulers.java [hunk 1]
```diff
@@ -78,9 +83,28 @@ public static PlatformSchedulers create(@NonNull final PlatformContext context,
...
+                        .withMetricsBuilder(model.metricsBuilder().withUnhandledTaskMetricEnabled(true))
+                        .build()
+                        .cast(),
+                // don't define a capacity for the postHashCollector, so that the postHashCollector will not apply
+                // backpressure to the hasher
+                model.schedulerBuilder("postHashCollector")
+                        .withType(TaskSchedulerType.SEQUENTIAL)
+                        .withOffRamp(hashingObjectCounter)
+                        .withFlushingEnabled(true)
+                        .withMetricsBuilder(model.metricsBuilder().withUnhandledTaskMetricEnabled(true))
```
Comment (cody-littley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/wiring/PlatformSchedulers.java:107]:
```suggestion
                        .withMetricsBuilder(model.metricsBuilder().withUnhandledTaskMetricEnabled(true))
                        withExternalBackPressure(true)
```

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/wiring/PlatformSchedulers.java [hunk 2]
```diff
@@ -78,9 +83,28 @@ public static PlatformSchedulers create(@NonNull final PlatformContext context,
...
         return new PlatformSchedulers(
                 model.schedulerBuilder("eventHasher")
-                        .withType(TaskSchedulerType.DIRECT)
+                        .withType(TaskSchedulerType.CONCURRENT)
+                        .withUnhandledTaskCapacity(config.eventHasherUnhandledCapacity())
+                        .withOnRamp(hashingObjectCounter)
+                        .withFlushingEnabled(true)
+                        .withMetricsBuilder(model.metricsBuilder().withUnhandledTaskMetricEnabled(true))
+                        .build()
+                        .cast(),
```
[the remark this answers is not in the document]
Comment (cody-littley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/wiring/PlatformSchedulers.java:100]: `withExternalBackPressure(true)` hints to the wiring model that there is back pressure on the wire. Should cause the line to be solid, and will allow detection of cyclical back pressure that includes the hasher.

**Comment by poulok** [comment]
This PR appears to implement the short term solution stated in the ticket. If that is the case, this PR should not close the ticket.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/wiring/components/PostHashCollectorWiring.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+ * Wiring object that allows for the staging of events that have been hashed, but haven't been passed further down the
+ * pipeline yet.
+ * <p>
+ * This wiring object is a workaround for the problems that concurrent schedulers currently have with pushing to
+ * components that apply backpressure.
+ *
+ * @param eventInput  the input wire for events that have been hashed
+ * @param eventOutput the output wire for events to be passed further along the pipeline
+ */
+public record PostHashCollectorWiring(
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/wiring/components/PostHashCollectorWiring.java:36]: I don't quite understand the purpose of this. Some more explanation would be useful for me

Comment (alittley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/wiring/components/PostHashCollectorWiring.java:47]: @lpetrovic05 I pushed an expanded description of the problem

**Review by edward-swirldslabs — APPROVED** [review]

**Review by edward-swirldslabs — APPROVED** [review]

**Review by cody-littley — APPROVED** [review]

**Review by cody-littley — APPROVED** [review]

**Pull request merged**